### PR TITLE
feat: improve session extend performance

### PR DESCRIPTION
**Pull request by aeneasr**
This patch improves the performance for extending session lifespans. Lifespan extension is tricky as it usually is part of the middleware of Ory Kratos consumers. As such, it is prone to transaction contention when we read and write to the same session row at the same time (and potentially multiple times).

To address this, we:

1. Introduce a locking mechanism on the row to reduce transaction contention;
2. Add a new feature flag which toggles returning 204 no content instead of 200 + session.

Be aware that all reads on the session table will have to wait for the transaction to commit before they return a value. This may cause long(er) response times on `/session/whoami` for sessions that are being extended at the same time.

BREAKING CHANGES: Going forward, the `/admin/session/.../extend` endpoint will return 204 no content for new Ory Network projects. We will deprecate returning 200 + session body in the future.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

* [ ] Add tests

### CODE REVIEW

**Comment by aeneasr** [comment]
You can play around yourself with table locking in this guide using `cockroach demo` as your DB: https://www.cockroachlabs.com/docs/stable/select-for-update#examples

**Review by jonas-jonas — APPROVED** [review]

File: persistence/sql/persister_session.go [hunk 1]
```diff
@@ -176,6 +179,50 @@ func (p *Persister) ListSessionsByIdentity(
...
+		if !s.CanBeRefreshed(ctx, p.r.Config()) {
+			return nil
+		}
+
+		s = s.Refresh(ctx, p.r.Config())
+		if _, err := tx.Where("id = ? AND nid = ?", sessionID, nid).UpdateQuery(s, "expires_at"); err != nil {
+			return sqlcon.HandleError(err)
+		}
+
+		trace.SpanFromContext(ctx).AddEvent(events.NewSessionLifespanExtended(ctx, s.ID, s.IdentityID, s.ExpiresAt))
```
Comment (alnr) [inline persistence/sql/persister_session.go:221]: This must be moved outside of the transaction into a deferred function call in order not to send duplicate events when the TX automatically retries (see `UpsertSession`).

File: persistence/sql/persister_session.go [hunk 2]
```diff
@@ -176,6 +179,50 @@ func (p *Persister) ListSessionsByIdentity(
...
+			// This is a special case for CockroachDB. If the row is locked, we return a specific error message.
+			if errors.Is(err, sqlcon.ErrNoRows) && tx.Dialect.Name() == dbal.DriverCockroachDB {
+				return errors.WithStack(herodot.ErrNotFound.WithReason("The session you are trying to extend is already being extended by another request or does not exist."))
+			}
+
+			return sqlcon.HandleError(err)
+		}
+
+		if !s.CanBeRefreshed(ctx, p.r.Config()) {
+			return nil
```
Comment (alnr) [inline persistence/sql/persister_session.go:213]: return `400`?

Comment (aeneasr) [reply persistence/sql/persister_session.go:213]: Would be a breaking change

File: persistence/sql/persister_session.go (hunk 1 above)
Comment (aeneasr) [reply persistence/sql/persister_session.go:221]: Nice catch - thanks!

**Pull request merged**